pre_commit/store.py

```diff
@@ -121,10 +122,7 @@ def _get_result():
                 return result
 
             logger.info('Initializing environment for {}.'.format(repo))
-
-            directory = tempfile.mkdtemp(prefix='repo', dir=self.directory)
-            with clean_path_on_failure(directory):
-                make_strategy(directory)
```

Comment:
I think you can put this back, cloning doesn't get involved for the `local` repositories (and this should happen anyway if the `make_strategy` raises)

Reply:
It is happening, I added "_perform_safe_clone" to make_local as well. The problem was that when you shallow clone (git init), you cannot reuse the directory for complete clone because of `fatal: destination path 'path' already exists and is not an empty directory.`. If I just created another directory in fallback, we would save invalid directory to the database so make_strategy had to return the directory that's going to be finally used.

I could change the make_strategy to a list and loop through it here. We wouldn't have to deal with error handling in make_strategy functions and multiple fallbacks would be possible (like the proposed 3). What do you think?

Reply:
another "equivalent" to full clone can reuse that though:

```bash
# non-shallow clone without using `git clone`
git init x
cd x
git remote add origin ...
git fetch origin  # this is all that changes between shallow / non-shallow clone
git checkout ...
```